model/pipeline/src/main/proto/metrics.proto

```diff
@@ -229,101 +215,127 @@ message MonitoringInfo {
+    //   - count: represents the number of values seen across all bundles
+    //   - sum: represents the total of the value across all bundles
+    //   - min: represents the smallest value seen across all bundles
+    //   - max: represents the largest value seen across all bundles
+    //
+    // Encoding: <count><sum><min><max>
+    //   - count: beam:coder:varint:v1
+    //   - sum:   beam:coder:double:v1
+    //   - min:   beam:coder:double:v1
+    //   - max:   beam:coder:double:v1
+    DISTRIBUTION_DOUBLE_TYPE = 3 [(org.apache.beam.model.pipeline.v1.beam_urn) =
+                                 "beam:metrics:distribution_int64:v1"];
+
+    // Represents the latest seen integer value. The timestamp is used to
+    // provide an "ordering" over multiple values to determine which is the
+    // latest.
+    //
+    // Encoding: <timestamp><value>
+    //   - timestamp: beam:coder:varint:v1     (milliseconds since epoch)
+    //   - value:     beam:coder:varint:v1
+    LATEST_INT64_TYPE = 4 [(org.apache.beam.model.pipeline.v1.beam_urn) =
+                               "beam:metrics:latest_int64:v1"];
+
+    // Represents the latest seen integer value. The timestamp is used to
+    // provide an "ordering" over multiple values to determine which is the
+    // latest.
+    //
+    // Encoding: <timestamp><value>
+    //   - timestamp: beam:coder:varint:v1     (milliseconds since epoch)
+    //   - value:     beam:coder:double:v1
+    LATEST_DOUBLE_TYPE = 5 [(org.apache.beam.model.pipeline.v1.beam_urn) =
+                           "beam:metrics:latest_int64:v1"];
+
+    // Represents the largest set of integer values seen across bundles.
+    //
+    // Encoding: <value1><value2>...<valueN>
+    //   - valueX: beam:coder:varint:v1
+    TOP_N_INT64_TYPE = 6 [(org.apache.beam.model.pipeline.v1.beam_urn) =
+                           "beam:metrics:top_n_int64:v1"];
+
+    // Represents the largest set of double values seen across bundles.
+    //
+    // Encoding: <value1><value2>...<valueN>
+    //   - valueX: beam:coder:double:v1
+    TOP_N_DOUBLE_TYPE = 7 [(org.apache.beam.model.pipeline.v1.beam_urn) =
+                            "beam:metrics:top_n_int64:v1"];
+
+    // Represents the smallest set of integer values seen across bundles.
+    //
+    // Encoding: <value1><value2>...<valueN>
+    //   - valueX: beam:coder:varint:v1
+    BOTTOM_N_INT64_TYPE = 8 [(org.apache.beam.model.pipeline.v1.beam_urn) =
+                          "beam:metrics:bottom_n_int64:v1"];
+
+    // Represents the smallest set of double values seen across bundles.
+    //
+    // Encoding: <value1><value2>...<valueN>
+    //   - valueX: beam:coder:double:v1
+    BOTTOM_N_DOUBLE_TYPE = 9 [(org.apache.beam.model.pipeline.v1.beam_urn) =
+                           "beam:metrics:bottom_n_int64:v1"];
+
+    // Encoding: <value1><value2>...<valueN>
+    //   - valueX: beam:coder:double:v1
+    PROGRESS_TYPE = 10 [(org.apache.beam.model.pipeline.v1.beam_urn) =
```

Comment:
I think progress should be a MonitoringInfoSpec URN and it should use either SUM_DOUBLE_TYPE or X_DOUBLE_TYPE.

It should be something which defscribes what each update is supposed to be, and how they should be aggregated together. I think that's required for each type, in order to all the use case of "worker just forwards these to a service to aggregate" As it may need to go pre aggregation of its own?

GAUGE_DOUBLE_TYPE? I am not sure what's appropriate here. But it probably doesn't make sense for progress to be its own type.

Reply:
This is the individual progress for each element and restriction pair for a splittable DoFn. The aggregation is interesting since if you sum across all **latest** values then that tells you the amount of completed work globally and the amount of known but not yet finished work globally. Note that this won't cover the amount of unknown work (e.g. elements and restrictions that have yet to be processed).

@robertwb / @ajamato What do you think about adding `ELEMENT` as labels (storing the encoded element and restriction) and then we could use `LATEST_DOUBLE_TYPE`?

Reply:
We don't want to report progress for each processed element, as this would become an unreasonably large value. We could report aggregated progress of all finished ones, plus partial progress of the in-flight ones, in which case the sum becomes a refinement of done-element-count. There's still no useful way to define cross-bundle aggregation (without resulting in an absolutely huge value), so I would lean towards this being its own type.